### PR TITLE
suggestion: support multiple db's at runtime

### PR DESCRIPTION
**Pull request by cylewitruk**
## Description

This PR is the skeleton of a _suggestion_ on how we could handle DB connections at runtime. Was something I played with over the weekend but when wiring things up in `main()` it started to feel more relevant (i.e. being able to run the signer locally without pgsql running).

## Changes

- Removes the `Error` associated type from `DbRead` and `DbWrite`: I don't really see the point of having them since these are traits which are contained within the signer binary themselves where we can use the crate's `Error` type. Also interferes with trait objects when the associated type differs.
- Introduces a new `DbReadWrite` trait which represents both a `DbRead` and `DbWrite`.
- Changes database access to use trait objects instead of static generics. This is unfortunately required to support multiple databases at runtime (i.e. if we want to allow in-memory/file-based for running the program locally vs. production which may choose pgsql).
- Centralizes the `DbReadWrite` impl to the `SignerContext` which is then responsible for handing out new handles to the underlying database.
  - This can be simplified if/when we implement sqlite, at least as an in-memory database in place of the current `vec`-based in-memory db. The issue is that the in-memory DB doesn't implement `sqlx::Pool` or `sqlx::Database` so we're forced to lift the abstraction up one level as it is today. Otherwise we could find a more elegant solution around the `sqlx` traits exclusively.
- My big hope is that we can abolish the current in-memory DB and use SQLite in-memory which will much more closely represent db access to pgsql etc.  The big issue with the current in-mem DB is that it won't accurately represent ANSI transactions (i.e. shared vs. exclusive read locks, exclusive write locks, page vs. row vs. column-level locks, etc.) nor will it fail on constraint violations. The current in-mem DB also uses a Mutex which will lock/block in situations where a traditional RDBMS won't.

## Checklist:

- [ ] I have performed a self-review of my code
- [ ] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by djordon — COMMENTED** [review]
I have a few notes:
1. We should brake this up into two PRs, and exclude the new `DbRead` functions from this one. But either way they need integration tests.
2. I don't see much benefit of using `async-trait` so we should remove it.
3. Consider making the db field in the `SignerContext` an enum instead of using dynamic dispatch. 
 
 
For point (3) implementing `DbRead` and `DbWrite` for such an enum involves a lot of mindless boiler plate, so if you considered it I can see why you looked the other way. And I don't think we can use `enum_dispatch` here because the functions for these traits return `impl Traits`. But yeah, let me know what you think.

File: signer/src/config.rs [hunk 1]
```diff
@@ -368,6 +384,19 @@ where
...
+fn db_endpoint_deserializer<'de, D>(deserializer: D) -> Result<url::Url, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    if s == IN_MEMORY_DB_SCHEME {
+        return url::Url::parse(&format!("{IN_MEMORY_DB_SCHEME}://"))
+            .map_err(serde::de::Error::custom);
+    }
+    s.parse().map_err(serde::de::Error::custom)
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/config.rs:397]: I'm open, I went a bit back and forth on this myself 🤷🏻‍♂️

File: signer/src/api/mod.rs [hunk 2]
```diff
@@ -4,16 +4,18 @@
...
+use crate::{config::Settings, storage::DbReadWrite};
 
 /// A struct with state data necessary for runtime operation.
-#[derive(Debug, Clone)]
-pub struct ApiState<S> {
+#[derive(Clone)]
+pub struct ApiState {
     /// For writing to the database.
-    pub db: S,
+    pub db: Arc<dyn DbReadWrite>,
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/api/mod.rs:18]: I had started with Box but ran into something that made me switch.. can't remember what off the top of my head... but Box _should_ work yeah

File: signer/src/storage/mod.rs [hunk 3]
```diff
@@ -10,243 +10,257 @@ pub mod in_memory;
...
+    fn as_read(self: Arc<Self>) -> Arc<dyn DbRead> {
+        self as Arc<dyn DbRead>
+    }
+    fn as_write(self: Arc<Self>) -> Arc<dyn DbWrite> {
+        self as Arc<dyn DbWrite>
+    }
+}
 
+/// Represents the ability to read data from the signer storage.
+#[async_trait]
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/mod.rs:43]: There was some issue with trait objects that async-trait resolved. I'm sure it's probably just something with how the return types are expressed, though.

File: signer/src/storage/postgres.rs [hunk 4]
```diff
@@ -740,12 +742,104 @@ impl super::DbRead for PgStore {
...
+            WHERE bitcoin_txid = $1 AND output_index = $2
+            ORDER BY created_at DESC;
+            "#,
+        )
+        .bind(Into::<BitcoinTxId>::into(outpoint.txid))
+        .bind(outpoint.vout as i64)
+        .fetch_optional(&self.0)
+        .await
+        .map_err(Error::SqlxQuery)
+    }
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/storage/postgres.rs:767]: Iirc this was to make your db tests in `new_block.rs` work with the new paradigm, but I'll have another look when I open up this code again 👍

File: signer/src/context.rs [hunk 5]
```diff
@@ -50,6 +54,33 @@ pub struct InnerSignerContext {
     /// for the duration of the program and is used to provide new senders
     /// and receivers for a [`TerminationHandle`].
     term_tx: tokio::sync::watch::Sender<bool>,
+    /// Handle to the database connection.
+    db: Arc<dyn DbReadWrite>,
```
[the remark this answers is not in the document]
Comment (cylewitruk) [reply signer/src/context.rs:58]: Yeah I did play with the enum route but it quickly cascaded into much larger changes/restruturing iirc (like impl DbRead/Write and routing all of the methods). I'm not against the enum approach, though

**Comment by cylewitruk** [comment]
@djordon i'm going to close this for now so it's not lingering, but we'll have it here if/when we want to pick up the discussion at a later time.

**Pull request closed without merging**